HSEARCH-1708 Throw an exception when sorting on a field not indexed

DavideD opened the pull request:
https://hibernate.atlassian.net/browse/HSEARCH-1708

hferentschik commented on engine/src/main/java/org/hibernate/search/query/engine/impl/HSQueryImpl.java:
```diff
@@ -670,7 +680,51 @@ else if ( this.sort != null && projection != null ) {
+	private void validateDistanceSortField(String field, Class<?> clazz) {
+		EntityIndexBinding indexBinding = extendedIntegrator.getIndexBinding( clazz );
+		PropertyMetadata propertyMetadata = indexBinding.getDocumentBuilder().getMetadata().getPropertyMetadataForProperty( field );
+		if ( propertyMetadata != null ) {
```
Would it not make more sense to enhance the metadata itself so that the `PropertyMetadata` can already tell you whether you are dealing with coordinates? It seems wrong to work with reflection this late in the game. Also it is quite expensive.

hferentschik replied:
If you check `DocumentFieldMetadata` it contains `isNumeric`. I would somehow expect that there is something like `isSpatial` as well.

Sanne commented on engine/src/main/java/org/hibernate/search/query/engine/impl/HSQueryImpl.java:
```diff
@@ -670,7 +680,51 @@ else if ( this.sort != null && projection != null ) {
 
+	private void validateDistanceSortField(String field, Class<?> clazz) {
+		EntityIndexBinding indexBinding = extendedIntegrator.getIndexBinding( clazz );
+		PropertyMetadata propertyMetadata = indexBinding.getDocumentBuilder().getMetadata().getPropertyMetadataForProperty( field );
```
this line here is problematic, the field name is not the same as the property name.

Sanne commented on engine/src/main/java/org/hibernate/search/query/engine/impl/HSQueryImpl.java:
```diff
@@ -639,6 +648,7 @@ else if ( Boolean.FALSE.equals( forceScoring ) ) {
 			);
 		}
 		else if ( this.sort != null && projection != null ) {
+			validate( sort, classesAndSubclasses );
```
Shouldn't we need to validate even if _sort_ is set but _projection_ is not?

hferentschik commented on engine/src/main/java/org/hibernate/search/query/engine/impl/HSQueryImpl.java:
```diff
@@ -670,7 +680,51 @@ else if ( this.sort != null && projection != null ) {
+			boolean indexedField = false;
+			for ( Class<?> clazz : classes ) {
+				IndexedTypeDescriptor indexedTypeDescriptor = extendedIntegrator.getIndexedTypeDescriptor( clazz );
+				FieldDescriptor fieldDescriptor = indexedTypeDescriptor.getIndexedField( field );
```
Here you are using the public metadata API opposed to the internal one (as you do later). `FieldDescriptor` seems to be the right spot to add additional metadata. There is for example already a `NumericFieldSettingsDescriptor`, maybe a `SpatialFieldSettingsDescriptor` would make sense. This way you might be able to skip accessing the internal metadata alltogether (obviously the internal metadata would need to change as well, since the public one is built from it, just saying you might be able to just use one type of API in this particular usecase)

Sanne commented on engine/src/main/java/org/hibernate/search/query/engine/impl/HSQueryImpl.java:
```diff
@@ -670,7 +680,51 @@ else if ( this.sort != null && projection != null ) {
+		PropertyMetadata propertyMetadata = indexBinding.getDocumentBuilder().getMetadata().getPropertyMetadataForProperty( field );
+		if ( propertyMetadata != null ) {
+			// The field is a property of the entity
+			XClass elementClass = propertyMetadata.getPropertyAccessor().getElementClass();
```
You shouldn't use the _XClass_ helpers here, as they will ignore other sources of metadata - for example the programmatic configuration.
Also, we shouldn't do reflection at runtime but somehow have information to validate this using our model.

DavideD commented:
Rebased and fixed.

I might have found a nicer solution: I validate the value when I read it; If it does not have a NumDocValues, it means that the field is invalid.

hferentschik commented on engine/src/main/java/org/hibernate/search/spatial/impl/SpatialIndexReader.java:
```diff
@@ -0,0 +1,45 @@
+ *
+ * @author Davide D'Alto
+ */
+public final class SpatialIndexReader {
```
We usually use `XYZHelper` to indicate helper classes with static methods. So I'd prefer `SpatialIndexReaderHelper` or just `SpatialHelper`. This avoids also the initial thought that we are dealing with a subclass of Lucene's `IndexReader` which was my first thought.

hferentschik commented:
> I might have found a nicer solution: I validate the value when I read it; If it does not have a NumDocValues, it means that the field is invalid. 

But can the `NumDocValue` not also just indicate another (non spatial) related field? A spatial fields is for sure a numeric field, but not every numeric field is use as spatial coordinate. I still think the metadata approach is more promising, especially since it adds more value to the (metadadata) API as well.

Sanne commented:
+1 to use the metadata.. you might not be able to read it if the index is empty anyway ;)

DavideD commented:
Remarks applied and rebased

gunnarmorling commented on engine/src/main/java/org/hibernate/search/spatial/impl/CoordinateHelper.java:
```diff
@@ -0,0 +1,33 @@
+	private CoordinateHelper() {
+	}
+
+	public static NumericDocValues loadCoordinateDocValues(AtomicReader atomicReader, String field) throws IOException {
```
What's the benefit of this over the direct method invocation?

DavideD replied:
right,  initially it was longer but after some clean up it became smaller.
I'll remove it

gunnarmorling commented on engine/src/main/java/org/hibernate/search/query/engine/impl/HSQueryImpl.java:
```diff
@@ -670,7 +678,32 @@ else if ( this.sort != null && projection != null ) {
+	}
 
+	private void validateDistanceSortField(ExtendedSearchIntegrator extendedIntegrator, SortField sortField) {
+		String field = sortField.getField();
```
This can be null for some sorts (e.g. by doc id).

DavideD replied:
I would like to add a test for this scenario. Do I just need to specify the document id as sort field?

gunnarmorling replied:
Yes, I think so.

gunnarmorling commented on engine/src/main/java/org/hibernate/search/query/engine/impl/HSQueryImpl.java:
```diff
@@ -670,7 +678,32 @@ else if ( this.sort != null && projection != null ) {
+				}
+			}
+		}
+		if ( !indexedField ) {
```
The method name `validateDistanceSortField()` is a bit too specific with this check here.

DavideD replied:
Not necessarily.
Take for example `SortTest#testCustomFieldComparatorAscendingSort`, this part in particular:
```Sort sort = new Sort( new SortField( "sum", new SumFieldComparatorSource() ) );```

"sum" is not an indexed field, but everything still works because `SumFieldComparatorSource` does not use the passed field name.

So, at the moment, I don't feel like imposing this validation for all sorts.
Probably, this is something for a separate issue.

gunnarmorling commented:
@DavideD, added some more minor comments. Looks alright apart from those.

hferentschik commented:
> Looks alright apart from those.

+1 That's what I had in mind solution wise. I also think it is a nice improvement for the metadata API. I guess you should still follow up on @gunnarmorling comment regarding the sort field being potentially `null`.

DavideD commented:
Remarks applied + few other changes:

* Null check for sortField
* Remove method from Coordinateshelper
* Squashed commits
* Log message print the SortField class

gunnarmorling commented:
Rebased and applied. Nice work, @DavideD!
